File: android/src/main/java/com/shingo/rtsp/RtspPlayer.java
package com.shingo.rtsp;

import android.content.Context;
import android.net.Uri;
import android.util.Log;
import android.view.View;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.videolan.libvlc.LibVLC;
import org.videolan.libvlc.Media;
import org.videolan.libvlc.MediaPlayer;
import org.videolan.libvlc.interfaces.IVLCVout;

import java.util.ArrayList;
import java.util.List;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.view.TextureRegistry;

public class RtspPlayer implements PlatformView, MethodChannel.MethodCallHandler {


    private LibVLC libVLC;
    private Media media;


    private VLCTextureView textureView;


    private MediaPlayer mediaPlayer;

    private List<String> options;
    private Context context;
    private int viewId;

    private MethodChannel methodChannel;

    public RtspPlayer(int viewId, Context context, BinaryMessenger binaryMessenger, TextureRegistry textureRegistry) {
        this.viewId = viewId;
        this.context = context;
        textureView = new VLCTextureView(context);
        methodChannel = new MethodChannel(binaryMessenger, "rtsp_player" + "_" + viewId);
        methodChannel.setMethodCallHandler(this);
    }


    public void initialize(List<String> options, String url, boolean autoPlay) {
        this.options = options;
        libVLC = new LibVLC(context, options);
        mediaPlayer = new MediaPlayer(libVLC);

        mediaPlayer.getVLCVout().setVideoView(textureView);
        mediaPlayer.getVLCVout().setWindowSize(textureView.getWidth(), textureView.getHeight());
        textureView.setMediaPlayer(mediaPlayer);
        mediaPlayer.setVideoTrackEnabled(true);

        mediaPlayer.getVLCVout().attachViews();
        media = new Media(libVLC, Uri.parse(url));
        for (int i = 0; i < options.size(); i++) {
            media.addOption(options.get(i));
        }
        media.setHWDecoderEnabled(true, false);
        mediaPlayer.setMedia(media);
        media.release();
        if (autoPlay) {
            play();
        }
    }


    void play() {
        if (mediaPlayer != null) {
            mediaPlayer.play();
        }
    }

    void pause() {
        if (mediaPlayer != null) {
            mediaPlayer.pause();
        }
    }

    void stop() {
        if (mediaPlayer != null) {
            mediaPlayer.stop();
        }
    }

    boolean isPlaying() {
        if (mediaPlayer == null) return false;
        return mediaPlayer.isPlaying();
    }


    @Nullable
    @Override
    public View getView() {
        return textureView;
    }

    @Override
    public void dispose() {
        release();
    }


    void release() {

        if (mediaPlayer != null) {
            mediaPlayer.stop();
            mediaPlayer.getVLCVout().detachViews();
            mediaPlayer.release();
        }
        if (libVLC != null) {
            libVLC.release();
            libVLC = null;
        }
        methodChannel = null;
        mediaPlayer = null;
        media = null;
        textureView = null;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        if ("initialize".equals(method)) {
            List<String> options = call.argument("options");
            String url = call.argument("url");
            boolean autoPlay = call.argument("autoPlay");

            initialize(options, url, autoPlay);
            result.success(true);
        } else if ("play".equals(method)) {
            play();
            result.success(true);
        } else if ("stop".equals(method)) {
            stop();
            result.success(true);
        } else if ("pause".equals(method)) {
            pause();
            result.success(true);
        } else if ("isPlaying".equals(method)) {
            result.success(isPlaying());
        }


    }
}

File: ios/Classes/RtspPlayer.swift
//
//  RtspPlayer.swift
//  flutter_rtsp_player
//
//  Created by shingohu on 2023/6/5.
//

import UIKit
import MobileVLCKit
import UIKit
import Flutter
import Foundation


public class RtspPlayer: NSObject ,FlutterPlatformView{
    var hostedView:UIView
    var vlcMediaPlayer: VLCMediaPlayer?
    
    
    
    
    init(frame: CGRect, viewId: Int64, messenger:FlutterBinaryMessenger) {
        
        self.hostedView = UIView(frame: frame)
        super.init()
        
        let methodChannel = FlutterMethodChannel(name: "rtsp_player_\(viewId)", binaryMessenger: messenger)
        
        methodChannel.setMethodCallHandler { (call, result) in
            
            let method = call.method
            
            if(method == "initialize"){
                let options = (call.arguments as! Dictionary<String, Any>)["options"] as! Array<String>
                
                let url =  (call.arguments as! Dictionary<String, Any>)["url"] as! String
                
                let autoPlay =  (call.arguments as! Dictionary<String, Any>)["autoPlay"] as! Bool
                
                
                self.initialize(options: options, url: url, autoPlay: autoPlay)
                
                result(true)
                
                
            }else if(method == "play"){
                self.play()
                result(true)
            }else if(method == "stop"){
                self.stop()
                result(true)
            }else if(method == "pause"){
                self.pause()
                result(true)
            }else if(method == "isPlaying"){
                result(self.isPlaying())
            }
            
            
        }
        
    }
    
    
    
    
    
    public func initialize(options:Array<String>,url:String,autoPlay:Bool){
        
        self.vlcMediaPlayer = VLCMediaPlayer.init(options: options)
        self.vlcMediaPlayer!.drawable = self.hostedView
        
        let media:VLCMedia = VLCMedia(url: URL(string: url)!);
        
        if(!options.isEmpty){
            for option in options {
                media.addOption(option)
            }
        }
        self.vlcMediaPlayer!.media = media
        if(autoPlay){
            self.vlcMediaPlayer!.play()
        }
        
    }
    
    
    
    public func play() {
        
        self.vlcMediaPlayer?.play()
    }
    
    public func pause() {
        
        self.vlcMediaPlayer?.pause()
    }
    
    public func stop() {
        
        self.vlcMediaPlayer?.stop()
    }
    
    public func isPlaying() -> Bool{
        if(self.vlcMediaPlayer != nil){
            return self.vlcMediaPlayer!.isPlaying 
        }
        return false
    }
    
    
    
    
    
    public func view() -> UIView {
        return hostedView
    }
    
    
}
